plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,945 @@
+                                arr[:, :, 0],
+                                arr[:, :, -1],
+                            )
+                        ]
+                    )
+                )
+
+                if edge_max > 1e-3 * np.max(arr):
+                    unit = grid.recognized_quantities[rq].unit
+                    warnings.warn(
+                        "Fields should go to zero at edges of grid to avoid "
+                        f"non-physical effects, but a value of {edge_max:.2E} {unit} was "
+                        f"found on the edge of the {rq} array. Consider applying a "
+                        "envelope function to force the fields at the edge to go to "
+                        "zero.",
+                        RuntimeWarning,
+                    )
+
+        # Validate inputs to the run function
+        # Sets is_synchronized_time property
+        self._validate_constructor_inputs(
+            termination_condition, save_routine, dt, dt_range, field_weighting
+        )
+
+        self.dt = dt.to(u.s).value if dt is not None else None
+
+        dt_range = [0, np.inf] * u.s if dt_range is None else dt_range
+        self.dt_range = dt_range.to(u.s).value
+
+        # Update the `tracker` attribute so that the stop condition & save routine can be used
+        termination_condition.tracker = self
+
+        if save_routine is not None:
+            save_routine.tracker = self
+
+        self.termination_condition = termination_condition
+        self.save_routine = save_routine
+
+    @property
+    def num_grids(self):
+        """The number of grids specified at instantiation."""
+        return len(self.grids)
+
+    def setup_adaptive_time_step(
+        self,
+        time_steps_per_gyroperiod: Optional[int] = 12,
+        Courant_parameter: Optional[float] = 0.5,
+    ):
+        """Set parameters for the adaptive time step candidates.
+
+        Parameters
+        ----------
+        time_steps_per_gyroperiod : int, optional
+            The number of subdivisions of the particle's gyration. The default is twelve.
+
+        Courant_parameter : float, optional
+            The Courant parameter is the fraction of a grid cell length by which to subdivide
+            the motion of the particle.
+
+
+        Notes
+        -----
+        Two candidates are calculated for the adaptive time step: a time step based on the gyroradius
+        of the particle and a time step based on the resolution of the grid. The candidate associated
+        with the gyroradius of the particle takes a ``time_steps_per_gyroperiod`` parameter that specifies
+        how many times the orbit of a gyrating particles will be subdivided. The other candidate,
+        associated with the spatial resolution of the grid object, calculates a time step using the time
+        it would take the fastest particle to cross some fraction of a grid cell length. This fraction is the Courant number.
+        """
+
```

Comment:
Should calling this method raise an exception if `self.dt` is not None, in other words if a fixed timestep is already set?

Reply:
I think this is what we're doing with the `not self.is_adaptive_time_step` condition?

Reply:
Ah, I see, yes that makes sense.